Sensors/SMCDellSensors/SMCDellSensors.cpp

```diff
@@ -0,0 +1,186 @@
+		//VirtualSMCAPI::addKey(KeyF0As(i), vsmcPlugin.data, VirtualSMCAPI::valueWithUint8(0, new F0As(i), SMC_KEY_ATTRIBUTE_WRITE | SMC_KEY_ATTRIBUTE_READ));
+		VirtualSMCAPI::addKey(KeyF0Mn(i), vsmcPlugin.data, VirtualSMCAPI::valueWithFp(0, SmcKeyTypeFpe2, new F0Mn(i), SMC_KEY_ATTRIBUTE_CONST | SMC_KEY_ATTRIBUTE_READ));
+		VirtualSMCAPI::addKey(KeyF0Mx(i), vsmcPlugin.data, VirtualSMCAPI::valueWithFp(0, SmcKeyTypeFpe2, new F0Mx(i), SMC_KEY_ATTRIBUTE_CONST | SMC_KEY_ATTRIBUTE_READ));
+		VirtualSMCAPI::addKey(KeyF0Md(i), vsmcPlugin.data, VirtualSMCAPI::valueWithUint8(0, new F0Md(i), SMC_KEY_ATTRIBUTE_WRITE | SMC_KEY_ATTRIBUTE_READ));
+		VirtualSMCAPI::addKey(KeyF0Tg(i), vsmcPlugin.data, VirtualSMCAPI::valueWithFp(0, SmcKeyTypeFpe2, new F0Tg(i), SMC_KEY_ATTRIBUTE_WRITE | SMC_KEY_ATTRIBUTE_READ));
+	}
+	VirtualSMCAPI::addKey(KeyFFRC, vsmcPlugin.data,
+		VirtualSMCAPI::valueWithUint16(0, new FFRC(), SMC_KEY_ATTRIBUTE_WRITE | SMC_KEY_ATTRIBUTE_READ));
+
+	OSArray* fanNames = OSDynamicCast(OSArray, getProperty("FanNames"));
+	char fan_name[DiagFunctionStrLen];
+
+	for (size_t i = 0; i < fanCount; i++) {
+		FanTypeDescStruct	desc;
+		FanInfo::SMMFanType type = SMIMonitor::getShared()->state.fanInfo[i].type;
+		if (type == FanInfo::Unsupported)
+			type = FanInfo::CPU;
+		snprintf(fan_name, DiagFunctionStrLen, "Fan %lu", i);
+		if (fanNames) {
+			OSString* name = OSDynamicCast(OSString, fanNames->getObject(type));
+			if (name)
+				lilu_os_strncpy(fan_name, name->getCStringNoCopy(), DiagFunctionStrLen);
+		}
+		lilu_os_strncpy(desc.strFunction, fan_name, DiagFunctionStrLen);
+		VirtualSMCAPI::addKey(KeyF0ID(i), vsmcPlugin.data, VirtualSMCAPI::valueWithData(
+			reinterpret_cast<const SMC_DATA *>(&desc), sizeof(desc), SmcKeyTypeFds, nullptr, SMC_KEY_ATTRIBUTE_CONST | SMC_KEY_ATTRIBUTE_READ));
+	}
+	
+	auto tempCount = min(SMIMonitor::getShared()->tempCount, MaxIndexCount);
+	for (size_t i = 0; i < tempCount; i++) {
+		TempInfo::SMMTempSensorType type = SMIMonitor::getShared()->state.tempInfo[i].type;
+		if (type <= TempInfo::Unsupported || type >= TempInfo::Last) {
+			DBGLOG("sdell", "Temp sensor type %d is unknown, auto assign value %d", type, SMIMonitor::getShared()->state.tempInfo[i].index);
+			type = static_cast<TempInfo::SMMTempSensorType>(SMIMonitor::getShared()->state.tempInfo[i].index);
+		}
+		
+		switch (type)
+		{
+		case TempInfo::CPU:
+			VirtualSMCAPI::addKey(KeyTC0P(0), vsmcPlugin.data, VirtualSMCAPI::valueWithSp(0, SmcKeyTypeSp78, new TC0P(i),
+				SMC_KEY_ATTRIBUTE_WRITE|SMC_KEY_ATTRIBUTE_READ));
+			break;
+		case TempInfo::GPU:
+			VirtualSMCAPI::addKey(KeyTG0P(0), vsmcPlugin.data, VirtualSMCAPI::valueWithSp(0, SmcKeyTypeSp78, new TG0P(i),
+				SMC_KEY_ATTRIBUTE_WRITE|SMC_KEY_ATTRIBUTE_READ));
+			break;
+		case TempInfo::Memory:
+			VirtualSMCAPI::addKey(KeyTm0P(0), vsmcPlugin.data, VirtualSMCAPI::valueWithSp(0, SmcKeyTypeSp78, new Tm0P(i),
+				SMC_KEY_ATTRIBUTE_WRITE|SMC_KEY_ATTRIBUTE_READ));
+			break;
```

Comment:
The same here, but more difficult to fix

Reply:
It is not an issue, there is a call before return:
qsort(const_cast<VirtualSMCKeyValue *>(vsmcPlugin.data.data()), vsmcPlugin.data.size(), sizeof(VirtualSMCKeyValue), VirtualSMCKeyValue::compare);